runtime/startup_byt.c

```diff
@@ -277,49 +277,67 @@ static char_os * read_section_to_os(int fd, struct exec_trailer *trail,
 
 static int parse_command_line(char_os **argv)
 {
-  int i, j;
+  int i, j, len, parsed;
 
   for(i = 1; argv[i] != NULL && argv[i][0] == '-'; i++) {
-    switch(argv[i][1]) {
-    case 't':
-      ++ caml_trace_level; /* ignored unless DEBUG mode */
-      break;
-    case 'v':
-      if (!strcmp_os (argv[i], T("-version"))){
-        printf ("%s\n", "The OCaml runtime, version " OCAML_VERSION_STRING);
-        exit (0);
-      }else if (!strcmp_os (argv[i], T("-vnum"))){
-        printf ("%s\n", OCAML_VERSION_STRING);
-        exit (0);
-      }else{
+    len = strlen_os(argv[i]);
+    parsed = 1;
+    if (len == 2) {
+      if (argv[i][1] == '-') {
+        i++;
+        break;
+      }
+      switch(argv[i][1]) {
+      case 't':
+        ++ caml_trace_level; /* ignored unless DEBUG mode */
+        break;
+      case 'v':
         caml_verb_gc = 0x001+0x004+0x008+0x010+0x020;
+        break;
+      case 'p':
+        for (j = 0; caml_names_of_builtin_cprim[j] != NULL; j++)
+          printf("%s\n", caml_names_of_builtin_cprim[j]);
+        exit(0);
+        break;
+      case 'b':
+        caml_record_backtrace(Val_true);
+        break;
+      case 'I':
+        if (argv[i + 1] != NULL) {
+          caml_ext_table_add(&caml_shared_libs_path, argv[i + 1]);
+          i++;
+        } else {
+          error("option '-I' needs an argument.");
+        }
+        break;
+      case 'm':
+        print_magic = 1;
+        break;
+      case 'M':
+        printf("%s\n", EXEC_MAGIC);
+        exit(0);
+        break;
+      default:
+        parsed = 0;
       }
-      break;
-    case 'p':
-      for (j = 0; caml_names_of_builtin_cprim[j] != NULL; j++)
-        printf("%s\n", caml_names_of_builtin_cprim[j]);
-      exit(0);
-      break;
-    case 'b':
-      caml_record_backtrace(Val_true);
-      break;
-    case 'I':
-      if (argv[i + 1] != NULL) {
-        caml_ext_table_add(&caml_shared_libs_path, argv[i + 1]);
-        i++;
+    } else {
+      if (!strcmp_os(argv[i], T("-version"))) {
+        printf("%s\n", "The OCaml runtime, version " OCAML_VERSION_STRING);
+        exit(0);
+      } else if (!strcmp_os(argv[i], T("-vnum"))) {
+        printf("%s\n", OCAML_VERSION_STRING);
+        exit(0);
+      } else if (argv[i][1] == 'I') {
+        caml_ext_table_add(&caml_shared_libs_path, &argv[i][2]);
```

Comment:
I'm no big fan of `-Idir` instead of `-I dir`.  Yes, C compilers traditionally support both forms, but none of the OCaml tools supports `-Idir`, if I'm not mistaken.

Reply:
flexlink does (granted, not one of the main tools!). I think I added this because I'd spotted that `-Ifoo` was erroneously parsed as `-I` and so went with the C compiler interpretation rather than keeping it an error message - I think given that I did know that flexlink can take them, I'd not realised that the driver didn't. I take it you'd be against updating the compiler driver to accept "glued" options too?

Reply:
Yes, flexlink implements "glued" options, and so does CompCert, because these tools try to look like C compilers and linkers.  But for other tools I prefer not to support glued options.  For one thing, the Arg stdlib module doesn't support them, so you need to roll up your own argument parsing code.  For another thing, a glued `-I` option makes it impossible to have other options starting with a capital `I`, and you never know when you'll need a `-Imaginary` or `-Italian` option :-)

Reply:
`-Itsgone` 🙂